### PR TITLE
Automated unit tests for functionalities

### LINKED ISSUE #35: Automated unit tests for functionalities
Now, as we are working in Test Driven Environment, writing unit tests is crucial for any software development.
1. Write unit tests for methods in .py files using pytest or unittest framework.
2. Create github workflow to run those unit tests for every pull-request raised.
3. Make sure that coverage should be above 90%

### PR DESCRIPTION
**Pull request by rubyruins**
# Description:

Added some tests using `unittest` to check the various functionalities if they are working as expected:
- Mocks data for 5 classes with 5 images each
- Checks if the directories for image paths are correctly set
- Checks if the classes for images are correctly found
- Checks if the filters are applied and found in the corresponding folders
- Checks if the processedData folder created correctly contains all the filtered images
- Checks if the sizes of train and test are split correctly
- Checks if the images are resized according to the dimensions given

Currently, I have put all the tests in a single file since parts of the code are dependant on each other. Tests are separated into different methods. I have if tested the various outputs (images/variables) are produced as expected and used `coverage.py` to check for code coverage. The coverage is above 90%.

Fixes #35 

## Type of change:

- [x] New feature (non-breaking change which adds functionality)
 
# Checklist:

- [x] I have performed a self-review of my own code.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.

# Screenshots / Video:

image

image

### CODE REVIEW

**Comment by rubyruins** [comment]
@chebroluharika, could you please give me suggestions if I can structure the code better. This is my first time unit testing so it may not be up to the mark.

File: tests/unit/Tests.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+		for j in range(5):
+			img = np.random.randint(0, 256, size = [150, 150, 3], dtype = np.uint8)
+			cv2.imwrite(f'./dataset_dir/data_dir/class{i}/class{i}_{j}.png', img)
+
+class testFlow2ML(unittest.TestCase):
+
+	create_data()
+	flow = Flow('dataset_dir', 'data_dir')
+	filters = ["median", "gaussian", "laplacian", "sobelx", "sobely"]
+	flow.applyFilters(filters)
```
Comment (chebroluharika) [inline tests/unit/Tests.py:24]: you should mock functions here and expect an output ratehr than calling direct function

**Comment by chebroluharika** [comment]
@rubyruins, you need to create test files for every python file. For example, take Data_Loader.py file, there are 4 functions inside it, where you need to write test cases for all them seperately.
Lets say, for getClasses(), call that actual function with params that you have created, according to it check output of that function using self.assertEqual(). 
Please get back to me if you have any more questions.

**Comment by chebroluharika** [comment]
@rubyruins If any of packages need to install to run unit tests, please mention them as well.
Create test_requirements.txt file and mention all those packages there.

**Review by chebroluharika — CHANGES_REQUESTED** [review]
you should mock functions here and expect an output rather than calling direct function

**Comment by rubyruins** [comment]
> Lets say, for getClasses(), call that actual function with params that you have created, according to it check output of that function using self.assertEqual().
> Please get back to me if you have any more questions.

So I think these functions are called after we create the Flow object first. So for testing each file, do we create the flow object for each of them and then test the functions? Since filters are dependant on flow for creating filtered images and dataloader is dependant on filters since it takes the filtered images and moves them to processedData before proceeding. Let us say we have the test_dataloader.py file, so first it will need the Flow object, then it will need the filters applied to them and only then  it can generate x_train, y_train and so on. So how would we proceed in those cases?

**Comment by chebroluharika** [comment]
@rubyruins I didn't understand for the creation of flow object to trigger dataloader class fuctions. It doesn't require if you are writing unit tests whereas they are needed for integration tests. For unit tests, you shoul create an instance for dataloader class and call functions of that class with that instance.
In unit tests, you should test each module seperately.

**Comment by rubyruins** [comment]
@chebroluharika, this the part I was talking about -

first we create the flow object:
`flow = Flow( 'dataset_dir' , 'data_dir' )`

Then on this object we are calling the applyFilters function,

`filters = ["median", "gaussian", "sobelx", "sobely"]`
`flow.applyFilters( filters )`

So without making the flow object, we will not be able to test the applyFilters function separately, right? 

Also again when we are splitting into test and train,
`(train_x, train_y, val_x, val_y) = flow.getDataset( img_dimensions, test_val_split )`

It takes the filtered images generated in above step and then moves them to processedData and creates the test and train from there. So without doing the above steps, how would we test the dataLoader function separately by passing arguments? 😅

**Comment by chebroluharika** [comment]
you need to mock all functions other than the function that you want to call

**Comment by chebroluharika** [comment]
Just a headsup, I resolved all @rubyruins queries in Discord channel

**Pull request merged**